src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_ObjectCreationExpression.cs

```diff
@@ -103,31 +103,48 @@ public override BoundNode VisitObjectCreationExpression(BoundObjectCreationExpre
 
         public override BoundNode VisitWithExpression(BoundWithExpression withExpr)
         {
-            RoslynDebug.AssertNotNull(withExpr.CloneMethod);
-            Debug.Assert(withExpr.CloneMethod.ParameterCount == 0);
             Debug.Assert(withExpr.Receiver.Type!.Equals(withExpr.Type, TypeCompareKind.ConsiderEverything));
 
             // for a with expression of the form
             //
             //      receiver with { P1 = e1, P2 = e2 }
             //
-            // we want to lower it to a call to the receiver's `Clone` method, then
+            // if the receiver is a struct, duplicate the value, then set the given struct properties:
+            //
+            //     var tmp = receiver;
+            //     tmp.P1 = e1;
+            //     tmp.P2 = e2;
+            //     tmp
+            //
+            // otherwise the receiver is a record class and we want to lower it to a call to its `Clone` method, then
             // set the given record properties. i.e.
             //
             //      var tmp = (ReceiverType)receiver.Clone();
             //      tmp.P1 = e1;
             //      tmp.P2 = e2;
             //      tmp
 
-            var cloneCall = _factory.Convert(
-                withExpr.Type,
-                _factory.Call(
-                    VisitExpression(withExpr.Receiver),
-                    withExpr.CloneMethod));
+            BoundExpression expression;
+
+            if (withExpr.Type.IsValueType)
+            {
+                expression = VisitExpression(withExpr.Receiver);
+            }
+            else
+            {
+                RoslynDebug.AssertNotNull(withExpr.CloneMethod);
```

Comment:
nit: We should be using an annotated Debug.Assert at this point, so it should be possible to just say `Debug.Assert(withExpr.CloneMethod is not null)`